prefixset_iter.go

```diff
@@ -0,0 +1,41 @@
+//go:build go1.23
+
+package netipds
+
+import (
+	"iter"
+	"net/netip"
+)
+
+// PrefixSeq returns an iterator over all prefixes in s.
+func (s *PrefixSet) PrefixSeq() iter.Seq[netip.Prefix] {
+	return func(yield func(netip.Prefix) bool) {
+		canYield := true
```

Comment:
I think `canYield` needs to be defined outside of the closure. Otherwise it'll be reset on each use of the iterator.

Reply:
Actually, we probably want `i` defined outside the iterator. A key feature of an iterator is that it remembers its position.

A user should be able `range` over some of the elements, then call `break` before exhausting the iterator, then pick up where they left off in a subsequent `range`.

Reply:
After changing the check in `checkYieldFalse` to `i > 0`, this implementation passes the test:
```go
// PrefixSeq returns an iterator over all prefixes in s.
func (s *PrefixSet) PrefixSeq() iter.Seq[netip.Prefix] {
	i := 0
	canYield := true
	return func(yield func(netip.Prefix) bool) {
		s.tree.walk(key{}, func(n *tree[bool]) bool {
			if canYield && n.hasEntry {
				canYield = yield(n.key.toPrefix())
				i++
			}
			canYield = canYield && i < s.size
			return !canYield
		})
	}
}
```

Reply:
> I think `canYield` needs to be defined outside of the closure. Otherwise it'll be reset on each use of the iterator.
>
> Actually, we probably want `i` defined outside the iterator. A key feature of an iterator is that it remembers its position.

This is not how Go's push iterators are supposed to work. In this case, you call` PrefixSeq()` to get an iterator, and for each use, by calling it directly or with `for range`, it starts from the beginning of the sequence. Some good examples are `slices.All` and friends: https://github.com/golang/go/blob/master/src/slices/iter.go. They keep the state of walking the slice inside the inner closure.

> A user should be able `range` over some of the elements, then call `break` before exhausting the iterator, then pick up where they left off in a subsequent `range`.

What you described would be a pull iterator, which can be converted from a push iterator with `iter.Pull` or `iter.Pull2`.

Reply:
Interesting! Shows what I know. 

After reading the docs, I see that they do accommodate the pattern I'm thinking of, [single-use iterators](https://pkg.go.dev/iter#hdr-Single_Use_Iterators), but clearly reusability is the idiom for in-memory data structures. I can also see how pull iterators also naturally support the pause-resume workflow I described. Thanks for those references.